controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,561 @@
-     * @return the list of segments active at a given timestamp.
-     * GetActiveSegments runs in O(n), where n is the total number of segments.
-     * It can be improved to O(k + logn), where k is the number of active segments at specified timestamp,
-     * using augmented interval tree or segment index..
-     *
-     * TODO: maintain a augmented interval tree or segment tree index
-     */
-    @Override
-    public synchronized CompletableFuture<List<Integer>> getActiveSegments(long timestamp) {
-        List<Integer> currentSegments = new ArrayList<>();
-        int i = 0;
-        while (i < segments.size() && timestamp >= segments.get(i).getStart()) {
-            if (segments.get(i).getEnd() >= timestamp) {
-                InMemorySegment segment = segments.get(i);
-                currentSegments.add(segment.getNumber());
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        synchronized (lock) {
+            if (historyTable == null) {
+                result.completeExceptionally(new DataNotFoundException("history"));
+            } else {
+                if (historyTable.getVersion().equals(updated.getVersion())) {
+                    historyTable = new Data<>(Arrays.copyOf(updated.getData(), updated.getData().length), updated.getVersion() + 1);
+                    result.complete(null);
+                } else {
+                    result.completeExceptionally(new WriteConflictException("segmentTable"));
+                }
             }
-            i++;
         }
-        return CompletableFuture.completedFuture(currentSegments);
+        return result;
     }
 
     @Override
-    public CompletableFuture<List<Segment>> startScale(List<Integer> sealedSegments, List<SimpleEntry<Double, Double>> keyRanges, long scaleTimestamp) {
-        Preconditions.checkNotNull(keyRanges);
-        Preconditions.checkArgument(keyRanges.size() > 0);
-
-        List<List<Integer>> predecessors = new ArrayList<>();
-        for (int i = 0; i < keyRanges.size(); i++) {
-            predecessors.add(new ArrayList<>());
+    CompletableFuture<Data<Integer>> getHistoryTable() {
+        if (this.historyTable == null) {
+            return FutureHelpers.failedFuture(new DataNotFoundException(getName()));
         }
 
-        int start = segments.size();
-
-        List<Segment> newSegments = new ArrayList<>();
-        // assign start times, numbers to new segments. Add them to segments list and current list.
-        for (int i = 0; i < keyRanges.size(); i++) {
-            int number = start + i;
-            InMemorySegment segment = new InMemorySegment(number, scaleTimestamp, Long.MAX_VALUE, keyRanges.get(i).getKey(), keyRanges.get(i).getValue(), InMemorySegment.Status.Active, new ArrayList<>(), predecessors.get(i));
-            newSegments.add(segment);
-            segments.add(segment);
-            currentSegments.add(number);
-        }
+        return CompletableFuture.completedFuture(copy(historyTable));
```

Comment:
I have put it under lock now. 
However, i dont agree that we need to protect 'reads' under lock in our use case. 
With an external store like zk we get the same behaviour where multiple steps in any operation are not atomic.
So it is the same case here with the in-memory store. If there is a race between reads and writes, and if reads are not lock protected, there are two cases really:
1. read gets old value from before write. 
2. read gets new value from after write. 
the caller is designed to handle both the cases. 
In our business logic only updates need CAS concurrency. 
So correctness is not broken by reading in a thread unsafe manner. 

But i have now wrapped reads under lock too so that we dont argue about this endlessly. However, lets start a discussion on this on slack in case you still think there is some incorrectness.